src/scipp/plt/model.py

```diff
@@ -0,0 +1,137 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+
+from .widgets.widget import WidgetView
+
+from typing import Tuple, Iterable
+from functools import partial
+
+
+class Node:
+
+    def __init__(self, func, name=None, views=None):
+        self.name = name
+        self.parent_name = None
+        self.dependency = None
+        self.func = func
+        self.views = list(views or [])
+
+    def send_notification(self, message):
+        for view in self.views:
+            view.notify({
+                "node_name": self.name,
+                "parent_name": self.parent_name,
+                "message": message
+            })
+
+    def request_data(self):
+        pipeline = []
+        node = self
+        while node.dependency is not None:
+            pipeline.append(node.func)
+            node = node.dependency
+        else:
+            pipeline.append(node.func)
+        pipeline = pipeline[::-1]
+        data = pipeline[0]()
+        for step in pipeline[1:]:
+            data = step(data)
+        return data
+
+    def add_view(self, view):
+        self.views.append(view)
+
+
+class Model:
+
+    def __init__(self, da):
+        self._name = da.name
+        self._nodes = {}
+        self["root"] = Node(name='root', func=lambda: da)
+
+    def __getitem__(self, name: str) -> Node:
+        return self._nodes[name]
+
+    def __setitem__(self, name: str, node: Node):
+        self._nodes[name] = node
+        node.parent_name = self._name
+        node.name = name
+
+    def items(self) -> Iterable[Tuple[str, Node]]:
+        yield from self._nodes.items()
+
+    def _children_of(self, name: str) -> Iterable[str]:
+        for node in self.nodes():
+            if node.dependency is not None and name == node.dependency.name:
+                yield node
+
+    def keys(self) -> Iterable[str]:
+        yield from self._nodes.keys()
+
+    def nodes(self) -> Iterable[str]:
+        yield from self._nodes.values()
+
+    def insert(self, name: str, node: Node, after: str):
+        assert after in self.keys()
+        self[name] = node
+        for child in self._children_of(after):
+            child.dependency = node
+        node.dependency = self[after]
+
+    def add(self, name: str, node: Node, after: str):
+        assert after in self.keys()
+        self[name] = node
```

Comment:
Can this corrupt the graph if there already is a node called `name`? That is, can it lead to dependencies between nodes that should not depend or could it create holes?

Reply:
You mean we should make sure that `name` doesn't exist in the graph before doing anything else? Good idea.

Reply:
Yes to your question